third_party/coro/coro.c

```diff
@@ -103,9 +103,30 @@ coro_init (void)
 
   coro_transfer (new_coro, create_coro);
 
-#if __GCC_HAVE_DWARF2_CFI_ASM && __amd64
-  asm (".cfi_undefined rip");
-#endif
+/*
+ * Call-chain ends here: we need to annotate this frame with CFI assembly to
+ * make unwinding stop here.
+ *
+ * References
+ * glibc:
+ * clone https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/x86_64/clone.S;h=31ac12da0cc08a934d514fed1de9eba1cb3e8ec5;hb=ebbb8c9f64c3486603ef4ccee4dd2a5574e41039
+ * start https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/start.S;h=9edd17b60cd54ec9eef11c76ab02322dcb5d057a;hb=5b736bc9b55115e67129e77db4de6cf193054cd2
+ * seastar: https://github.com/scylladb/seastar/blob/d27bf8b5a14e5b9e9c9df18fd1306489b651aa42/src/core/thread.cc#L278-L293
+ * julia: https://github.com/JuliaLang/julia/blob/2e2b1d2ad50fe12999cbded0b5acd3f0a36ec8c5/src/julia_internal.h#L90-L106
+ * android: https://cs.android.com/android/platform/superproject/+/master:bionic/libc/platform/bionic/macros.h;l=52-60;drc=2528dab7419a63f57fe20027886ba7dd3857aba8
+ */
+#ifdef HAVE_CFI_ASM
+  /*
+   * See https://github.com/libunwind/libunwind/blob/ec171c9ba7ea3abb2a1383cee2988a7abd483a1f/src/x86_64/Gstep.c#L91-L92
+   * and https://github.com/libunwind/libunwind/blob/ec171c9ba7ea3abb2a1383cee2988a7abd483a1f/src/dwarf/Gparser.c#L877.
+   */
+  __asm__(".cfi_undefined rip\n"
+	  ".cfi_undefined rbp\n"
+	  ".cfi_return_column rbp\n");
```

Comment:
Shall we celar `%rbp` here also, as it is mentioned in `thread_start` and `__start`?

Reply:
I think this is redundant: see other references on the list — AFAIC, unwinders ignore the call-chain ABI anyways.

Reply:
Not quite. [Libunwind](https://github.com/dropbox/libunwind/blob/16bf4e5e498c7fc528256843a4a724edc2753ffd/src/x86_64/Gstep.c#L153) tells me that in some cases `%rbp` value is a decision-making for the unwinder.

Reply:
> Shall we celar %rbp here also, as it is mentioned in thread_start and __start?

`thread_start` and `__start` are written in assembly, while we use inline assembly — clearing rbp here would clobber the compiler's register.

Reply:
> Not quite. [Libunwind](https://github.com/dropbox/libunwind/blob/16bf4e5e498c7fc528256843a4a724edc2753ffd/src/x86_64/Gstep.c#L153) tells me that in some cases `%rbp` value is a decision-making for the unwinder.

Please see the `else if`[ branch](https://github.com/dropbox/libunwind/blob/16bf4e5e498c7fc528256843a4a724edc2753ffd/src/x86_64/Gstep.c#L136-L140) above: the branch you indicate will never be reached if `rbp` is undefined:
```c
else if (DWARF_IS_NULL_LOC (c->dwarf.loc[RBP]))
{
    for (i = 0; i < DWARF_NUM_PRESERVED_REGS; ++i)
      c->dwarf.loc[i] = DWARF_NULL_LOC;
 }
```

Reply:
Honestly, I see no reason for discussion here. As it is said in [System V Application Binary Interface spec](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) (see section 3.4.1 Initial Stack and Register State)

| **%rbp** The content of this register is unspecified at process initialization time, but the __user code should mark the deepest stack frame by setting the frame pointer to zero__.

So, please follow ABI here, not some referenced implementations.

Looks like discussion is resolved by the following changes:

```c
  "\txorq %%rbp, %%rbp\n"
```

Reply:
@Buristan, yes — I forgot to report here, that we agreed on this change with @sergos.

Reply:
Fixed: put coro’s target function call into assembly and zeroed out rbp, as discussed.

Reply:
fine!